### PR TITLE
Use `#[rustc_inherit_overflow_checks]` instead of Add::add etc.

### PR DESCRIPTION
**Pull request by m-ou-se**
See https://github.com/rust-lang/rust/issues/81721

### CODE REVIEW

File: library/core/src/iter/traits/accum.rs [hunk 1]
```diff
@@ -43,28 +42,44 @@ macro_rules! integer_sum_product {
         #[$attr]
         impl Sum for $a {
             fn sum<I: Iterator<Item=Self>>(iter: I) -> Self {
-                iter.fold($zero, Add::add)
+                iter.fold(
+                    $zero,
+                    #[rustc_inherit_overflow_checks]
+                    |a, b| a + b,
```
Comment (eddyb) [inline library/core/src/iter/traits/accum.rs:48]: Hah, does this work? I half-expected it not to.

Comment (m-ou-se) [reply library/core/src/iter/traits/accum.rs:48]: Haha, this was [*your suggestion*](https://github.com/rust-lang/rust/issues/81721#issuecomment-772786852).. But yeah, it compiles. I think it works too, but I should check if it actually gets tested for this behaviour. :)

Comment (eddyb) [reply library/core/src/iter/traits/accum.rs:48]: Yeah, I was going off having seen `#[inline(never)}`, IIRC, on closures, but I wasn't sure it would work out of the box.

**Comment by m-ou-se** [comment]
Okay this looks good, no regression in performance, and even a few percent improvement for cranelift.

To do: Verify that it actually works the same as before. I'm not sure if we test all these cases for ther release/debug-mode panic behaviour.

**Comment by m-ou-se** [comment]
> To do: Verify that it actually works the same as before. I'm not sure if we test all these cases for ther release/debug-mode panic behaviour.

Manually verified that it behaves the same. Looks like there's also already tests for these in `src/test/ui/iterators/iter-*-overflow-debug.rs`.

File: library/core/src/iter/traits/accum.rs [hunk 2]
```diff
@@ -37,34 +36,49 @@ pub trait Product<A = Self>: Sized {
...
-// N.B., explicitly use Add and Mul here to inherit overflow checks
 macro_rules! integer_sum_product {
     (@impls $zero:expr, $one:expr, #[$attr:meta], $($a:ty)*) => ($(
         #[$attr]
         impl Sum for $a {
             fn sum<I: Iterator<Item=Self>>(iter: I) -> Self {
-                iter.fold($zero, Add::add)
+                iter.fold(
+                    $zero,
+                    #[rustc_inherit_overflow_checks]
```
Comment (Mark-Simulacrum) [inline library/core/src/iter/traits/accum.rs:46]: Hm, so I initially thought "this seems wrong, don't we need it on the `fn sum` too? But it looks like in practice inherit is a bit wrong - this attribute just forces the MIR to contain overflow checks in all cases, and then when we codegen (potentially in a downstream crate) they'll get removed. So I think we're good -- not sure what a better name for the attribute is, but inherit definitely feels a bit off.

Comment (m-ou-se) [reply library/core/src/iter/traits/accum.rs:46]: Yeah, the `inherit` here is about crates rather than functions, in some sense.

Comment (eddyb) [reply library/core/src/iter/traits/accum.rs:46]: Yeah, "inherit" refers to the crate it's codegen'd in. So it's more "defer"? Or we can go more explicit really:

`#[rustc_use_overflow_checks_settings_from_user_crate]`

We could also make the attribute warn/error if building the MIR doesn't end up making a decision based on it, because that's what usually happens if it's applied to the wrong thing (e.g. a function doing `iter.fold(...)` instead of a closure).
